File: src/linkedList.cpp
#include "linkedList.h"
#include <iostream>
#include <cstring>
#include <fstream>

using namespace std;

// Структура узла списка
struct ListNode {
    string data;
    ListNode* next;
};

class LinkedList {
private:
    ListNode* head; // Указатель на голову списка
    ListNode* tail; // Указатель на хвост списка

public:
    // Инициализация списка
    void init() {
        head = nullptr;
        tail = nullptr;
    }

    // Добавление нового элемента в начало
    void addToHead(const string& value) {
        ListNode* newNode = new ListNode{value, head};
        head = newNode;
        if (tail == nullptr) {
            tail = head;
        }
    }

    // Добавление нового элемента в конец
    void addToTail(const string& value) {
        ListNode* newNode = new ListNode{value, nullptr};
        if (tail != nullptr) {
            tail->next = newNode;
        }
        tail = newNode;
        if (head == nullptr) {
            head = tail;
        }
    }

    // Удаление элемента с головы
    void removeFromHead() {
        if (head == nullptr) {
            return;
        }
        ListNode* temp = head;
        head = head->next;
        if (head == nullptr) {
            tail = nullptr;
        }
        delete temp;
    }

    // Удаление элемента с хвоста
    void removeFromTail() {
        if (tail == nullptr) {
            return;
        }
        if (head == tail) {
            delete head;
            head = nullptr;
            tail = nullptr;
            return;
        }
        ListNode* temp = head;
        while (temp->next != tail) {
            temp = temp->next;
        }
        delete tail;
        tail = temp;
        tail->next = nullptr;
    }

    // Удаление узла по значению
    void removeByValue(const string& value) {
        if (head == nullptr) {
            return;
        }
        if (head->data == value) {
            removeFromHead();
            return;
        }
        ListNode* temp = head;
        while (temp->next != nullptr) {
            if (temp->next->data == value) {
                ListNode* nodeToRemove = temp->next;
                temp->next = temp->next->next;
                if (nodeToRemove == tail) {
                    tail = temp;
                }
                delete nodeToRemove;
                return;
            }
            temp = temp->next;
        }
    }

    // Поиск элемента по значению
    bool search(const string& value) {
        ListNode* temp = head;
        while (temp != nullptr) {
            if (temp->data == value) {
                return true;
            }
            temp = temp->next;
        }
        return false;
    }

    // Вывод всех элементов списка
    void print() {
        ListNode* temp = head;
        while (temp != nullptr) {
            cout << temp->data << " ";
            temp = temp->next;
        }
        cout << endl;
    }

    // Очистка списка
    void destroy() {
        while (head != nullptr) {
            removeFromHead();
        }
    }

    // Загрузка элементов из файла
    void loadFromFile(const string& fileName) {
        ifstream file(fileName);
        string value;
        while (file >> value) {
            addToTail(value);
        }
        file.close();
    }

    // Сохранение элементов списка в файл
    void saveToFile(const string& fileName) {
        ofstream file(fileName);
        ListNode* temp = head;
        while (temp != nullptr) {
            file << temp->data << endl;
            temp = temp->next;
        }
        file.close();
    }
};

// Функция запуска работы со списком, обработка команд из аргументов командной строки
void runLinkedList(int argc, char* argv[]) {
    LinkedList list;
    list.init();

    string fileName;
    string query;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--file") == 0 && i + 1 < argc) {
            fileName = argv[i + 1];
            i++;
        } else if (strcmp(argv[i], "--query") == 0 && i + 1 < argc) {
            query = argv[i + 1];
            i++;
        }
    }

    list.loadFromFile(fileName);

    string command;
    size_t pos = query.find(' ');
    if (pos != string::npos) {
        command = query.substr(0, pos);
        query = query.substr(pos + 1);
    } else {
        command = query;
    }

    if (command == "LPUSH") {
        list.addToHead(query);  // Добавление элемента в начало списка
        list.saveToFile(fileName);
    } else if (command == "LAPPEND") {
        list.addToTail(query);  // Добавление элемента в конец списка
        list.saveToFile(fileName);
    } else if (command == "LREMOVEHEAD") {
        list.removeFromHead();  // Удаление элемента с головы списка
        list.saveToFile(fileName);
    } else if (command == "LREMOVETAIL") {
        list.removeFromTail();  // Удаление элемента с конца списка
        list.saveToFile(fileName);
    } else if (command == "LREMOVE") {
        list.removeByValue(query);  // Удаление элемента по значению
        list.saveToFile(fileName);
    } else if (command == "LSEARCH") {
        cout << (list.search(query) ? "true" : "false") << endl;    // Поиск элемента по значению
    } else if (command == "LPRINT") {
        list.print();
    }

    list.destroy();
}

File: Go/src/linkedList.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// ListNode — узел списка
type ListNode struct {
	data string
	next *ListNode
}

// LinkedList — односвязный список
type LinkedList struct {
	head *ListNode
	tail *ListNode
}

func (l *LinkedList) Init() {
	l.head = nil
	l.tail = nil
}

func (l *LinkedList) AddToHead(value string) {
	newNode := &ListNode{data: value, next: l.head}
	l.head = newNode
	if l.tail == nil {
		l.tail = l.head
	}
}

func (l *LinkedList) AddToTail(value string) {
	newNode := &ListNode{data: value, next: nil}
	if l.tail != nil {
		l.tail.next = newNode
	}
	l.tail = newNode
	if l.head == nil {
		l.head = l.tail
	}
}

func (l *LinkedList) RemoveFromHead() {
	if l.head == nil {
		return
	}
	l.head = l.head.next
	if l.head == nil {
		l.tail = nil
	}
}

func (l *LinkedList) RemoveFromTail() {
	if l.tail == nil {
		return
	}
	if l.head == l.tail {
		l.head = nil
		l.tail = nil
		return
	}
	cur := l.head
	for cur.next != nil && cur.next != l.tail {
		cur = cur.next
	}
	cur.next = nil
	l.tail = cur
}

func (l *LinkedList) RemoveByValue(value string) {
	if l.head == nil {
		return
	}
	if l.head.data == value {
		l.RemoveFromHead()
		return
	}
	prev := l.head
	cur := l.head.next
	for cur != nil {
		if cur.data == value {
			prev.next = cur.next
			if cur == l.tail {
				l.tail = prev
			}
			return
		}
		prev = cur
		cur = cur.next
	}
}

func (l *LinkedList) Search(value string) bool {
	cur := l.head
	for cur != nil {
		if cur.data == value {
			return true
		}
		cur = cur.next
	}
	return false
}

func (l *LinkedList) Print() {
	first := true
	for cur := l.head; cur != nil; cur = cur.next {
		if !first {
			fmt.Print(" ")
		}
		first = false
		fmt.Print(cur.data)
	}
	fmt.Println()
}

func (l *LinkedList) Destroy() {
	l.head = nil
	l.tail = nil
}

// LoadFromFile — читает слова из файла и добавляет в хвост (поведение как у C++ >>)
func (l *LinkedList) LoadFromFile(fileName string) error {
	f, err := os.Open(fileName)
	if err != nil {
		return err
	}
	defer f.Close()
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		l.AddToTail(scanner.Text())
	}
	return scanner.Err()
}

func (l *LinkedList) SaveToFile(fileName string) error {
	f, err := os.Create(fileName)
	if err != nil {
		return err
	}
	defer f.Close()
	w := bufio.NewWriter(f)
	for cur := l.head; cur != nil; cur = cur.next {
		_, _ = w.WriteString(cur.data + "\n")
	}
	return w.Flush()
}

// runLinkedList — обрабатывает команды списка
func runLinkedList(fileName string, fullQuery string) {
	list := LinkedList{}
	list.Init()
	_ = list.LoadFromFile(fileName) // если файла нет — получим пустой список

	parts := strings.SplitN(fullQuery, " ", 2)
	command := parts[0]
	arg := ""
	if len(parts) > 1 {
		arg = parts[1]
	}

	switch command {
	case "LPUSH":
		list.AddToHead(arg)
		_ = list.SaveToFile(fileName)
	case "LAPPEND":
		list.AddToTail(arg)
		_ = list.SaveToFile(fileName)
	case "LREMOVEHEAD":
		list.RemoveFromHead()
		_ = list.SaveToFile(fileName)
	case "LREMOVETAIL":
		list.RemoveFromTail()
		_ = list.SaveToFile(fileName)
	case "LREMOVE":
		list.RemoveByValue(arg)
		_ = list.SaveToFile(fileName)
	case "LSEARCH":
		if list.Search(arg) {
			fmt.Println("true")
		} else {
			fmt.Println("false")
		}
	case "LPRINT":
		list.Print()
	default:
		fmt.Fprintln(os.Stderr, "Неизвестная команда:", command)
	}

	list.Destroy()
}
